name: CI

on:
  pull_request:
    types: [opened, synchronize, reopened]
    branches:
      - master
      - v1.0

jobs:
  Lint:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout branch
        uses: actions/checkout@v2

      - name: Setup Python 3.6 üêç
        uses: actions/setup-python@v1.1.1
        with:
          python-version: 3.6
      
      - name: Set up the project
        run: |
          pip install poetry safety
          poetry config virtualenvs.in-project true
          make setup

      - name: Check the code quality
        run: make check-code-quality

      - name: Check if the code is correctly typed
        run: make check-types

      - name: Check for vulnerabilities in dependencies
        run: make check-dependencies

  Tests:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version: [3.6, 3.7, 3.8]

    steps:
      - name: Checkout branch
        uses: actions/checkout@v2

      - name: Setup Python ${{ matrix.python-version }} üêç
        uses: actions/setup-python@v1.1.1
        with:
          python-version: ${{ matrix.python-version }}

      - name: Cache pip
        uses: actions/cache@v1
        with:
          path: ~/.cache/pip # This path is specific to Ubuntu
          # Look to see if there is a cache hit for the corresponding requirements file
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-

      - name: Set up the project
        run: |
          pip install poetry
          poetry config virtualenvs.in-project true
          make setup
          
      - name: Run the test suite
        run: make test

      - name: Upload coverage to Codecov
        if: matrix.python-version == 3.7
        uses: codecov/codecov-action@v1
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          file: ./coverage.xml
          flags: unittests
          name: codecov-umbrella
          yml: codecov.yml
          fail_ci_if_error: true
